Adds `DirectBeamEnabled` attribute

mockoocy opened the pull request:
Adds `DirectBeamEnabled` attribute.
Also implements some of it's behaviours. That is:
1. it disallows moving beamstop out of `BEAM` position when fast shutter is open
2. it disallows opening fast shutter when beamstop is not at `BEAM` position.

For these 2 points above, I've made a change to `Attribute` class. 
It takes a callback, that is supposed to validate / transform value when setting value. 
This allows the checks to happen when the value is assigned. It could also allow stuff like normalizing `omega` angle to [0,360) range, during assignment (but that's for a different PR :smiley_cat:)

As the `DirectBeamEnabled` is not yet implemented in MXCuBE (and changing it directly would also not be possible), I tested it using this small script below.
```python
```python
#!/usr/bin/env python3
import socket
import sys

HOST = "b-micromax-md3-pc"
PORT = 9001


def usage():
    print(f"{sys.argv[0]} <attribute_name> <value>")
    sys.exit()


def get_command():
    if len(sys.argv) != 3:
        usage()
    attribute_name, value = sys.argv[1:]
    return f"\x02WRTE {attribute_name} {value}\x03".encode()


def main():
    command = get_command()
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.connect((HOST, PORT))
        s.sendall(command)
        while True:
            print(s.recv(1024))

main()
```

mockoocy committed:
Adds DirectBeamEnabled attribute

This attribute is present in the real MD3.
It enables / disables moving diffractometer components in a way that
allows beam hitting detector directly (that is, without encountering
fast shutter / beamstop)

mockoocy committed:
Restrict some attribute's values when direct beam is disabled

Modifies the Attribute class, so that it also takes a callback function
in the constructor. It is a mechanism that allows performing value
validation / transformation before a value is set.

Makes it impossible to simultaneously have fast shutter open and
beamstop out of BEAM position when DirectBeamEnabled=False.

elmjag commented:
Looks promising!

However, we should also emulate correct replies when operations are disallowed. I have tested with the real MD3Up.

Given that `DirectBeamEnabled` is set to `False`. Here are the MD3Up replies.

**Beamstop at PARK position, Fast Shutter is closed**

When sending:

` b"\x02WRTE FastShutterIsOpen True\x03"`

the reply is:

`b'\x02ERR:Cannot change value to: true\x03'`

**Beamstop at BEAM position, Fast Shutter is open**

When sending:

`b"\x02EXEC setBeamstopPosition PARK\t\x03"`

the reply is:

`b'\x02ERR:Invalid value\x03'`

(Don't ask me the logic for these replies, but this is what we have).

elmjag commented:
Btw, I quickly whopped together the script below for testing. It have 'commands' for enabling/disabling direct beam, moving beam stop and opening and closing fast shutter. Perhaps you also find it useful:

```
#!/usr/bin/env python3
import socket
import sys

HOST = "b-micromax-md3-pc"
PORT = 9001

COMMANDS = {
    "back_on": b"\x02WRTE BackLightIsOn True\x03",
    "back_off": b"\x02WRTE BackLightIsOn False\x03",
    "front_on": b"\x02WRTE FrontLightIsOn True\x03",
    "front_off": b"\x02WRTE FrontLightIsOn False\x03",
    "dbeam_on": b"\x02WRTE DirectBeamEnabled True\x03",
    "dbeam_off": b"\x02WRTE DirectBeamEnabled False\x03",
    "fshutter_open": b"\x02WRTE FastShutterIsOpen True\x03",
    "fshutter_close": b"\x02WRTE FastShutterIsOpen False\x03",
    "bstop_beam": b"\x02EXEC setBeamstopPosition BEAM\t\x03",
    "bstop_park": b"\x02EXEC setBeamstopPosition PARK\t\x03",
}


def usage():
    print(f"{sys.argv[0]} <command>\n\ncommands:")
    for cmd in COMMANDS:
        print(f"    {cmd}")

    sys.exit()


def get_command():
    if len(sys.argv) != 2:
        usage()

    command = sys.argv[1].lower()
    if command not in COMMANDS:
        usage()

    return COMMANDS[command]


def main():
    command = get_command()
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.connect((HOST, PORT))
        s.sendall(command)
        while True:
            print(s.recv(1024))


main()
```

mockoocy committed:
Uses error messages from a real MD3

It works by propagating `DisallowedState` exceptions from methods that
execute task / attribute setters and dispatching them at the "outermost"
layer (that is in `handle_write, handle_exec` methods).

elmjag commented:
Very good!

I have changed the code formatting a bit, squashed all commits and merged this. Good job!
